### DISCUSSION TITLE
CC: Tweaked 1.104 for MC 1.19.4

### DISCUSSION THREAD

**Discussion by SquidDev** [cc-tweaked/CC-Tweaked#1400] [Announcements]
While a couple of weeks later than I was hoping, happy to announce CC: Tweaked 1.104, the first release for Minecraft 1.19.4. While quite a lot has changed, I'm afraid most of it is pretty boring internal stuff, which always makes it hard to write these release notes!

## Register to vote!
If you're in the UK, and of an age to do so, make sure you're registered for the upcoming local elections (May 4th). The deadline is April 17th, but make sure you do it well before then - it only takes 5 minutes.

Remember you will now need to take photographic ID with you. Yes, it's bloody stupid.

## Forge _and_ Fabric
As mentioned in [the 1.102/1.103 release notes](https://github.com/cc-tweaked/CC-Tweaked/discussions/1330), multi-loader support was planned for Minecraft 1.19.4. Thank you to everyone who has helped test this, and of course for the CC: Restitched team for maintaining the original Fabric port for so long.

## Re-licensing
Back in Feburary I posted [my plans for the partial relicensing of CC: Tweaked](https://github.com/cc-tweaked/CC-Tweaked/issues/1339). I've now started work on this and, thanks to the generous permission of many past contributors, have relicensed about half the codebase under the MPL 2.0.

This doesn't really have any practical difference on how you consume the mod as a whole, as the distributed jars are still licensed under the CCPL. But hopefully this should make it easier to reuse some of our code in other projects.

## Cobalt changes
We updated to Cobalt 0.7  (our Lua runtime) this release. This shouldn't have any practical impact on how you use the mod, but some things worth mentioning:

 - Coroutines _never_ require threads to yield any more, meaning we no longer need the coroutine thread pool. [I wrote a blog post on how we achieved that](https://squiddev.cc/2023/03/29/coroutines-and-bytecode.html), which you should check out if you enjoy silly compiler things.
 - `table` methods (such as `table.insert`) and `ipairs` now obey metamethods like `__index`.
 - The way we handle timeouts was entirely rewritten, which should make it a little more reliable.

---

As always, thank you for everyone who has contributed to this release!

**Comment by MCJack123** [comment]
> * `multishell` now hides the implementation details of its terminal redirect from programs.

I'm a little bit concerned about the implications of this change. I've seen a lot of people using `term.current().getLine()` to capture the contents of the screen, and I've always warned them that it may not exist and not to rely on it. Previously, this was only the case when running on a normal computer (or one with `bios.use_multishell` disabled), but now it will never exist. I'm hoping this doesn't break too much stuff (though if it does, __I warned you!__).

**Reply by Lupus590** [reply]
The fix will be as simple as using a window right? It shouldn't be too difficult for people to patch programs.

**Reply by SquidDev** [reply]
Yeah, this change was made explicitly so that people would stop doing that. While Hyrum's law is always at play here, this is one case where I think everyone has been very clear that `term.current().getLine` was not recommended in any way (and indeed, was entirely broken when using `bios.use_multishell` or alternative shells like `mbs`).

**Reply by SquidDev** [reply]
For some actual motivation for the issue. As part of #1245, I'm thinking about with replacing the usage of `window` in multishell with something more specialised. Nothing is really set in stone yet, but it seemed sensible to make this change in advance.
